src/consensus/aft/raft.h

```diff
@@ -1183,10 +1164,56 @@ namespace aft
       {
         if (i <= state->last_idx)
         {
-          // If the current entry has already been deserialised, skip the
-          // payload for that entry
-          ledger->skip_entry(data, size);
-          continue;
+          // NB: This is only safe as long as AppendEntries only contain a
+          // single term. If they cover multiple terms, then we would need to at
+          // least partially deserialise each entry to establish what term it is
+          // in (or report the terms in the header)
+          static_assert(
+            max_terms_per_append_entries == 1,
+            "AppendEntries rollback logic assumes single term");
+          const auto incoming_term = r.term_of_idx;
+          const auto local_term = state->view_history.view_at(i);
+          if (incoming_term != local_term)
+          {
+            if (is_new_follower)
+            {
+              auto rollback_level = i - 1;
+              RAFT_DEBUG_FMT(
+                "New follower received AppendEntries with conflict. Incoming "
+                "entry {}.{} conflicts with local {}.{}. Rolling back to {}.",
+                incoming_term,
+                i,
+                local_term,
+                i,
+                rollback_level);
+              LOG_ROLLBACK_FMT(
+                "Dropping conflicting branch. Rolling back {} entries, "
+                "beginning with {}.{}.",
+                state->last_idx - rollback_level,
+                local_term,
+                i);
+              rollback(rollback_level);
```

Comment:
This is one of the key changes - moving this `rollback` from outside of the loop (whenever we see a conflict-y AppendEntries, even if it contains no heartbeat, we immediately rollback) to here (when considering each entry in an AppendEntries, if we see one which conflicts, we rollback).

Note that this is now the _only_ path where we reset `is_new_follower` - we're allowed _one conflicting AE which causes rollback_ per-term, where previously we'd only obey it if it was the _first_ AE we received (and eagerly reset `is_new_follower` otherwise).